Implemented build restore without blocking UI thread

jainaashish opened the pull request:
This PR avoid blocking UI thread to block build while running onbuild restore. So instead of using restore on `onBuildEvent` from dte, it now runs it with `UpdateSolution_QueryDelayFirstUpdateAction` and delay build until restore is in progress.

So, now VS won't be freezed or unresponsive during build restore and improve user experience as well as performance.

Fixes https://github.com/NuGet/Home/issues/4986
Resolves NuGet/Home#4457.

@rrelyea @DoRonMotter

Commit:
Implemented build restore without blocking UI thread

emgarten commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -158,9 +189,30 @@ public void UpdateSolution_QueryDelayFirstUpdateAction(out int pfDelay)
                     pfDelay = 1;
                 }
+                // check if no build restore is running, then start a new one
+                else if (!_issRestoreRunning)
```
``_is``

emgarten reviewed (approved):
Minor spelling issue. Looks great overall. I went through this with Ashish offline.

alpaix commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -158,9 +189,30 @@ public void UpdateSolution_QueryDelayFirstUpdateAction(out int pfDelay)
+                    OnBuildRestore();
+
+                    // delay build until restore is running
+                    pfDelay = 1;
```
This introduces unnecessary delay when `OnBuildRestore` no-ops. For instance, when performing Clean only, or `ShouldRestoreOnBuild` returns `false`.

alpaix commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -90,6 +88,11 @@ public sealed class RestoreManagerPackage : AsyncPackage
+                _solutionBuildManager3 = (IVsSolutionBuildManager3)await GetServiceAsync(typeof(SVsSolutionBuildManager));
+                Assumes.Present(_solutionBuildManager3);
+
+                _solutionBuildManager3.AdviseUpdateSolutionEvents3(updateSolutionEvent, out _updateSolutionEventsCookie3);
```
Validate the return code.

alpaix commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -176,6 +228,16 @@ public void UpdateSolution_QueryDelayFirstUpdateAction(out int pfDelay)
 
             public void OnActiveProjectCfgChangeBatchEnd() { }
+
+            public int OnBeforeActiveSolutionCfgChange(IVsCfg pOldActiveSlnCfg, IVsCfg pNewActiveSlnCfg)
```
Why is this implementation needed?

alpaix reviewed (changes requested):
This change implements non-trivial state machine. My major concern is validate all edge cases are covered. For instance, one issue I found when build is cancelled during restore running. In this case the state machine would be broken.

alpaix committed:
Addressed PR feedback

Resolves NuGet/Home#4457.

- Refactored event listener into a standalone class for testability.
- Changed event model for better control over restore operation.
Now it handles four key events `UpdateSolution_Begin`,
`UpdateSolution_QueryDelayFirstUpdateAction`, `UpdateSolution_Cancel`,
`UpdateSolution_Done`.
- Added tests to cover all scenarios.

alpaix commented:
🔔 @alpaix @emgarten Addressed the feedback, added tests.

alpaix reviewed (approved):
LGTM

emgarten commented on test/TestUtilities/Test.Utility/Test.Utility.csproj:
```diff
@@ -46,6 +62,7 @@
   <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard1.6' ">
     <Compile Remove="PackageManagement\*.cs" />
     <Compile Remove="ProjectManagement\*.cs" />
+    <Compile Remove="Threading\*.cs" />
```
👍

emgarten reviewed (approved):
LGTM
